### ISSUE TITLE
False positive with variable key in Mapping with Optional value

### ISSUE THREAD

**Issue by aphedges** [python/mypy#10163]
**Bug Report**

When using a mapping with an `Optional` value type, accessing a value using a variable key instead of a literal key prevents a `None` check from working properly.

**To Reproduce**

1. Put the following code in `test.py`:

```python
from typing import Mapping, Optional

FOO: str = "foo"

def func(text: str) -> None:
    _ = text

data: Mapping[str, Optional[str]] = {FOO: "bar"}

if data[FOO] is not None:
    func(data[FOO])  # Line 11

if data["foo"] is not None:
    func(data["foo"])
```
2. Run `mypy test.py`

**Expected Behavior**

mypy should not output any errors. As the above example shows, using a string literal instead of a variable containing a string does not cause any errors, even though the code is otherwise identical. The ["Optional types and the None type"](https://mypy.readthedocs.io/en/stable/kinds_of_types.html#strict-optional) section of the documents says "supported checks for guarding against a `None` value include `if x is not None`", so this code should work.

**Actual Behavior**

mypy outputs the following:

```console
test.py:11: error: Argument 1 to "func" has incompatible type "Optional[str]"; expected "str"  [arg-type]
```

**Your Environment**

- Mypy version used: `0.812`
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): `show_error_codes = True`
- Python version used: `3.7.9`
- Operating system and version: macOS 10.15.7 (19H524)

**Comment by JukkaL** [comment]
Yeah, mypy should probably be able to figure this out, at least if `FOO` is final:

```py
FOO: Final = "foo"
```

This already works if you use a literal string as the key (e.g. `data["foo"]`).

**Comment by aphedges** [comment]
Using `Final` (from `typing_extensions`) for both `FOO` and `data` was one of the first things I tried, but I didn't include it in my example because it didn't change the error message at all.
